### PR TITLE
Non-sink endpoint characteristics

### PR DESCRIPTION
**Pull request by tiferet**
This PR simply follows the chain of logic that determines that an endpoint is not a sink [here](https://github.com/github/codeql/blob/387e57546bf7352f7c1cfe781daa1a3799b7063e/javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointData.qll#L75), and converts all the reasoning into `EndpointCharacteristic`s. These new characteristics will get used in [the next PR](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2098) to select endpoints as negative training samples.

Timing checks:
- ✅ KPI timing experiment: https://github.com/github/codeql-dca-main/issues/8438
- ✅ The local runtime of `endpoint_large_scale/ExtractEndpointData` and `endpoint_large_scale/ExtractEndpointDataTraining` are not impacted

closes https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2097

### CODE REVIEW

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:129]: For now, I've created a hierarchy of characteristics that closely reflects the hierarchy of classes and predicates in the old implementation. I'm happy to change this hierarchy if there's a different one that makes more sense. This would require JS expertise I don't have, so for now I followed the old implementation.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:153]: All the places where I've set a confidence are a fairly arbitrary interpretation of "NotASink" and "LikelyNotASink". For now the numbers are set so as to replicate the behavior of the old code, but these confidence settings can later be revised by someone with JS expertise (e.g. Stephan) 😄

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:48]: Future PRs will add even more EndpointCharacteristics, replacing the current endpoint filters. We might want to break these across multiple files, but that will probably better be done once we know how the hierarchy looks and are closer to the final list of characteristics 🤷

**Review by kaeluka — COMMENTED** [review]
I've left some comments :)

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:138]: These are some graphs I made for my own comprehension, but they may be useful for others.

## Main

This is the hierarchy we have on `main` (but with the new `OtherModeledArgumentCharacteristic` added in):

```mermaid
graph TD;
    string-->EndpointC;
    EndpointC-->DomBasedXssSinkC;
    EndpointC-->TaintedPathSinkC;
    EndpointC-->SqlInjectionSinkC;
    EndpointC-->NosqlInjectionSinkC;
    EndpointC-->OtherModeledArgumentC*;
```

## PR
This is the to-be-added sub-hierarchy. You'll have to zoom in, or read the markdown of this comment, or own a huge screen :)

In addition to what I said above,
* I think it is probably wrong that `LikelyNotASinkCharacteristic` and `NotASinkCharacteristic` are subclasses of `OtherModeledArgumentCharacteristic`. For example, there should be the possibility for building not-a-sinks that are NOT modeled arguments.
* I don't think that all builtin functions are also modeled, therefore `ArgumentToBuiltinFunctionCharacteristic` should not extend `OtherModeledArgument`.
* 

```mermaid
graph TD;
    OtherModeledArgumentC*-->ArgToBuiltinFuncC*;
    OtherModeledArgumentC*-->NotASinkC*;
    OtherModeledArgumentC*-->LikelyNotASinkC*;
    ArgToBuiltinFuncC*-->ArgToArrayC*;
    ArgToBuiltinFuncC*-->ArgToBuiltinGlobalVarRefC*;
    ArgToBuiltinFuncC*-->ConstantReceiverC*;
    ArgToBuiltinFuncC*-->BuiltinCallNameC*;
    NotASinkC*-->Lodash*;
    NotASinkC*-->JQArgC*;
    NotASinkC*-->ClientRequestC*;
    NotASinkC*-->PromiseDefC*;
    NotASinkC*-->CryptoKeyC*;
    NotASinkC*-->LoggerMethodC*;
    NotASinkC*-->TimeoutC*;
    NotASinkC*-->ReceiverStorageC*;
    NotASinkC*-->StringStartsWithC*;
    NotASinkC*-->StringEndsWithC*;
    NotASinkC*-->StringRegExpTestC*;
    NotASinkC*-->EvRegistrationC*;
    NotASinkC*-->EvDispatchC*;
    NotASinkC*-->MemberCandTC*;
    NotASinkC*-->FSAccessC*;
    NotASinkC*-->DBAccessC*;
    NotASinkC*-->DomC*;
    NotASinkC*-->NextFuncCallC*;
    NotASinkC*-->DojoRequireC*;
    NotASinkC*-->B64ManipC*;
    LikelyNotASinkC*-->ArgToArrayC*;
    LikelyNotASinkC*-->ArgToBuiltinGlobalVarRefC*;
```

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:34]: The class is the int returned by `getEncoding`. [`NegativeType`](https://github.com/kaeluka/codeql/blob/ac14b6d68549e62571134e70c334bef82f826ffc/javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointTypes.qll#L25-L26) is class 0.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:138]: AIUI, the hierarchy in the PR directly reproduces the hierarchy on main. Is that untrue? It's entirely possible I messed something up, but that was my intention.

As for improving on the hierarchy that currently exists on main, I'm 💯 happy to do so. In this PR I simply attempted to reproduce main with the new framework, because I don't have the JS knowledge to make assertions such as "not all builtin functions are also modeled". You _do_ have that knowledge, so I think you're the right person to make improvements to the hierarchy. I would suggest doing that in a subsequent PR, that you author and someone else with JS knowledge reviews, so as to separate the switch to the new framework from substantive changes to the way we categorize endpoints.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:146]: > > This class reads a bit strange and I think this is worth hashing out.
> > This abstract class wants to state that any of the 20 classes extending it will have a `getImplications` predicate that states that any of the endpoints satisfying their `getEndpoints` predicate are a high-confidence-isPositive-`Negatives`.
> > I'm surprised that the class does NOT state that these endpoints are also:
> > 
> > * high-confidence-_isNegative_-NosqlInjectionSinkType
> > * high-confidence-_isNegative_-SqlInjectionSinkType
> > * high-confidence-_isNegative_-XssSinkType
> > * high-confidence-_isNegative_-TaintedPathSinkType
> 
> That would be redundant. If we already know with high confidence that these endpoints are not sinks of _any_ type, why would we need to specify also that they are not NoSQL sinks, not SQL sinks, etc.? Also, that would require adding every new sink type we want to boost to this list.
> 
> The reason we have the ability to specify something like `high-confidence-_isNegative_-NosqlInjectionSinkType` is for characteristics that are inherently sink-type specific, such as our current query-specific endpoint filters. If I have good reason to say a certain set of endpoints are not NoSQL sinks, but I don't know anything about whether or not they're, say, Xss sinks, then I'll use `high-confidence-_isNegative_-NosqlInjectionSinkType`.
> 
> > Or maybe it should state instead that these endpoints are:
> > 
> > * _low-confidence_-isPositive-NosqlInjectionSinkType
> > * _low-confidence_-isPositive-SqlInjectionSinkType
> > * _low-confidence_-isPositive-XssSinkType
> > * _low-confidence_-isPositive-TaintedPathSinkType
> 
> This would be wrong. `_low-confidence_-isPositive-NosqlInjectionSinkType` means that I think this endpoint _is_ a NoSQL sink, but I'm not very confident about that assertion. See [the definition of isPositiveIndicator](https://github.com/github/codeql/blob/ac14b6d68549e62571134e70c334bef82f826ffc/javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll#L36): `isPositiveIndicator: If true, this characteristic indicates that this endpoint _is_ a member of the class; if false, it indicates that it _isn't_ a member of the class.`
> 
> Low-confidence indicators may come in handy for speeding up inference: If we think an endpoint is not very likely to be a sink, it may not be worth featurizing and scoring it. For the training set, OTOH, we probably won't look at characteristics that give only a weak indication, because we want to be fairly certain that most of our training samples have a correct label.
> 
> > Or should it state both?
> > I'm a bit puzzled.. I don't have a good intuition about this confidence mechanism; that there's a boolean IN ADDITION to a confidence score seems to add to my confusion (perhaps replacing that by a confidence score [-1, 1] would help some?).
> 
> See the [definition of the confidence](https://github.com/github/codeql/blob/ac14b6d68549e62571134e70c334bef82f826ffc/javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll#L38), and especially the example about what a low-confidence implication tells us: `A confidence near zero means this characteristic is a very weak indicator of whether or not the endpoint belongs to the class.` In practice we'll probably never bother writing EndpointCharacteristics that have confidence very close to zero, because characteristics that tell us almost nothing aren't useful. As mentioned above, though, we might use characteristics with confidence well below 1 to filter out unlikely candidates at inference time, to speed up code scanning, at the risk of maybe missing out on a few sinks our models might have otherwise detected.
> 
> Is there anything I could do to clarify the doc string more? If this was confusing for you, it might confuse others down the line as well.
> 
> Also, if it's still confusing, maybe we should sync about it 🙂

Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:146]: The permalinks are now fixed -- sorry about that!

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll]: I've tried to make this clearer in 243980ef73d14e2099e6d4fb98abc7a7904f5e66. Please LMK if it's still confusing 🙏

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:139]: > I think that the whole class-hierarchy starting at `OtherModeledArgumentCharacteristic` could be merged into one `private class OtherModeledArgumentCharacteristic` that has a `getImplications` implementation that contains the disjunction of all the classes' `getImplications` that you're proposing to add here. If not that, at least some classes in the sub-hierarchy could be merged in this fashion.
> 
> Or is there a specific reason to use such a big class hierarchy?
> 
> Any opinion, @henrymercer?

Did you mean `getEndpoints` rather than `getImplications`?

Potentially we could get rid of `OtherModeledArgumentCharacteristic` and `ArgumentToBuiltinFunctionCharacteristic` entirely, because at the moment they don't implement any predicates, and all they do is preserve the hierarchy that existed in the old code, with no actual implications (that is, if we removed them, it would change nothing about the current selection of endpoints). Again I would prefer to do this in a subsequent PR that simplifies/improves the existing hierarchy, rather than in this one.

We could also possibly get rid of all the classes that inherit from `NotASinkCharacteristic`, and implement them as a single disjunction of all the classes' `getEndpoints`, as you suggested, pairing one string description with each selection of endpoints. We could then do the same for all the classes that inherit from `LikelyNotASinkCharacteristic`. That would make the code shorter, but IMHO it would make it less clear. Right now we have one class per characteristic, which translates directly into one string in the list of characteristics we'll assign to each training endpoint. If a couple of these are really similar we can merge them together, but overall I think we're better off with one `EndpointCharacteristic` class per logical characteristic. This makes it very easy to experiment with changes to one characteristic without accidentally affecting anything else, and is a fairly critical advantage of the new design.

Incidentally, the list of characteristics _does_ get used by the training code. Suppose I have 1000 training samples with `ClientRequest` but only 10 samples with `PromiseDefinition`. If we train the model naively, to minimize the loss it will try to learn the patterns behind `ClientRequest`, but won't care too much whether or not it learns the patterns behind `PromiseDefinition`. To prevent this, I give higher weight in the loss function to samples that have `PromiseDefinition` in their list of characteristics. This pushes the model to learn these less common patterns as well as the most common ones. When I first implemented this change in the modeling code it had significant impact, though I haven't experimented with turning it off any time lately, so I don't know how big the impact is these days.

In the future I think we might also want to subsample negative samples that have very common characteristics, so that we can achieve a smaller training set without sacrificing the diversity of our negative samples.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll]: Implemented in dbcdc2209e556484408971a0188e63fbed2377ae

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:159]: After our sync discussion yesterday, I'm satisfied with the direction this is going in. I'm not 100% convinced that the first formulation is really redundant **but now is not the best time to figure that out - time will tell**.

If we'd have found out that any of these additional relations would've been necessary, we'd have needed to take a step back, and I'm glad we didn't :)

I'm resolving this conversation.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (henrymercer) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:151]: From a quick look, this PR introduces a new class hierarchy based on the implementation of the predicates in `CoreKnowledge`.  This would concern me a little in terms of BDD usage, and I don't think it's the right long term solution, as Stephan points out.

That said, if performance is OK, then I don't have quealms with introducing a temporary complex class hierarchy, as long as we remove it by the end of this refactoring.

**Comment by kaeluka** [comment]
OK, I made another review pass; Only one discussion is open 🥳

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll
Comment (tiferet) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:151]: I'm happy to push a commit, but can you tell me exactly what hierarchy you want? Incidentally, when I spoke of a subsequent PR, I meant immediately following this one, as soon as this one is merged. It's a trivial change to make, though, as long as I'm clear on your desired hierarchy. If it's easier than describing it in English, you can even create a branch off of this one to implement the change you want there. Rather than merging it into this PR, I'll copy it manually so you're not a committer 😊. If possible, I'd love to get this PR merged today, before my long weekend.

Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:151]: After our conversation yesterday, I'm convinced this won't get lost or forgotten. It's fine to do it in a later PR! 👍

**Review by kaeluka — APPROVED** [review]
👍

**Pull request merged**
